subprojects/native/src/main/java/org/gradle/internal/nativeintegration/jvm/Java9Detector.java

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.nativeintegration.jvm;
+
+import org.gradle.api.JavaVersion;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * Detects whether or not Gradle is being run on Java 9, and logs a warning message the first time it's detected.
+ */
+public class Java9Detector {
+    private static final Logger LOGGER = LoggerFactory.getLogger(Java9Detector.class);
+    private static final String SKIPPING_ENVIRONMENT = "Running with Java 9, the Gradle Daemon will not have the Gradle Client's environment variables.";
+    private AtomicBoolean hasWarned = new AtomicBoolean(false);
+
+    public boolean isJava9() {
+        if (JavaVersion.current().isJava9Compatible()) {
+            if (!hasWarned.getAndSet(true)) {
```

Comment:
Sorry to be dense here, but I don't really understand the motivation for either of these requested changes.

1. It seems like it should be the responsibility of the implementations of `ProcessEnvironment` to only call the `setEnvironmentVariable()` and `removeEnvironmentVariable()` methods if there are changes in the environment. But, that's not the current behavior. When we call `maybeSetEnvironment()` the implementation unconditionally attempts to set those environment variables. Are you suggesting that we need to add the overhead of reading them, checking that they are different, and then writing them only if they are different just to avoid the possible confusion of logging this message? I think I'd prefer to just change the wording of the message to make it clear that we won't be setting the daemon's environment to match the client's should it differ.

2. `SingleMessageLogger` doesn't extend `Logger`. So, I cannot use it to emit a simple warning-level log message. All of the `nagUser*` methods on `SingleMessageLogger` currently seem to be tailored for use-cases which don't match what we're trying to convey to the user. For example, we're not deprecating anything or discontinuing an API. So, I don't see how using the `SingleMessageLogger` is helpful. Are you suggesting that I should add a new method to that logger? Maybe `warnUserOfUnexpectedBehavior(String message)` or something like that?